scipy/fft/_basic.py

```diff
@@ -1,26 +1,58 @@
-from scipy._lib.uarray import generate_multimethod, Dispatchable
+from scipy._lib._array_api import array_namespace, is_numpy
+from . import _basic_uarray
 import numpy as np
 
 
-def _x_replacer(args, kwargs, dispatchables):
-    """
-    uarray argument replacer to replace the transform input array (``x``)
-    """
-    if len(args) > 0:
-        return (dispatchables[0],) + args[1:], kwargs
-    kw = kwargs.copy()
-    kw['x'] = dispatchables[0]
-    return args, kw
+def arg_err_msg(param):
+    return f'Providing {param!r} is only supported for numpy arrays.'
 
 
-def _dispatch(func):
-    """
-    Function annotation that creates a uarray multimethod from the function
-    """
-    return generate_multimethod(func, _x_replacer, domain="numpy.scipy.fft")
+def _validate_fft_args(overwrite_x, workers, plan, norm):
+    if overwrite_x:
+        raise ValueError(arg_err_msg("overwrite_x"))
+    if workers is not None:
+        raise ValueError(arg_err_msg("workers"))
+    if plan is not None:
+        raise ValueError(arg_err_msg("plan"))
+    if norm is None:
+        norm = 'backward'
+    return norm
+
+
+def _execute_1D(func_str, uarray_func, x, n, axis, norm,
+                overwrite_x, workers, plan):
+    xp = array_namespace(x)
+    if is_numpy(xp):
+        return uarray_func(x, n=n, axis=axis, norm=norm,
+                           overwrite_x=overwrite_x, workers=workers, plan=plan)
+
+    norm = _validate_fft_args(overwrite_x, workers, plan, norm)
+    if hasattr(xp, 'fft'):
+        xp_func = getattr(xp.fft, func_str)
+        return xp_func(x, n=n, axis=axis, norm=norm)
+
+    x = np.asarray(x)
+    y = uarray_func(x, n=n, axis=axis, norm=norm)
+    return xp.asarray(y)
+
+
+def _execute_nD(func_str, uarray_func, x, s, axes, norm, 
+                overwrite_x, workers, plan):
+    xp = array_namespace(x)
+    if is_numpy(xp):
+        return uarray_func(x, s=s, axes=axes, norm=norm,
+                           overwrite_x=overwrite_x, workers=workers, plan=plan)
+
+    norm = _validate_fft_args(overwrite_x, workers, plan, norm)
+    if hasattr(xp, 'fft'):
+        xp_func = getattr(xp.fft, func_str)
+        return xp_func(x, s=s, axes=axes, norm=norm)
+
+    x = np.asarray(x)
+    y = uarray_func(x, s=s, axes=axes, norm=norm)
+    return xp.asarray(y)
```

Comment:
Is it a design decision to keep separate "dispatchers" for 1D vs n-D here? I was able to combine them and have the test results unchanged locally using the diff below the fold. I don't know if there's a separate testing mechanism with `uarray` I'm not using though, that needs i.e., a keyword-only call signature or something? I guess the same pattern happens in `_realtransforms.py` as well, though those are even smaller funcs.

<details>

```diff
--- a/scipy/fft/_basic.py
+++ b/scipy/fft/_basic.py
@@ -19,37 +19,20 @@ def _validate_fft_args(overwrite_x, workers, plan, norm):
     return norm
 
 
-def _execute_1D(func_str, uarray_func, x, n, axis, norm,
-                overwrite_x, workers, plan):
-    xp = array_namespace(x)
-    if is_numpy(xp):
-        return uarray_func(x, n=n, axis=axis, norm=norm,
-                           overwrite_x=overwrite_x, workers=workers, plan=plan)
-
-    norm = _validate_fft_args(overwrite_x, workers, plan, norm)
-    if hasattr(xp, 'fft'):
-        xp_func = getattr(xp.fft, func_str)
-        return xp_func(x, n=n, axis=axis, norm=norm)
-
-    x = np.asarray(x)
-    y = uarray_func(x, n=n, axis=axis, norm=norm)
-    return xp.asarray(y)
-
-
 def _execute_nD(func_str, uarray_func, x, s, axes, norm, 
                 overwrite_x, workers, plan):
     xp = array_namespace(x)
     if is_numpy(xp):
-        return uarray_func(x, s=s, axes=axes, norm=norm,
+        return uarray_func(x, s, axes, norm=norm,
                            overwrite_x=overwrite_x, workers=workers, plan=plan)
 
     norm = _validate_fft_args(overwrite_x, workers, plan, norm)
     if hasattr(xp, 'fft'):
         xp_func = getattr(xp.fft, func_str)
-        return xp_func(x, s=s, axes=axes, norm=norm)
+        return xp_func(x, s, axes, norm=norm)
 
     x = np.asarray(x)
-    y = uarray_func(x, s=s, axes=axes, norm=norm)
+    y = uarray_func(x, s, axes, norm=norm)
     return xp.asarray(y)
 
 
@@ -193,7 +176,7 @@ def fft(x, n=None, axis=-1, norm=None, overwrite_x=False, workers=None, *,
     >>> plt.show()
 
     """
-    return _execute_1D('fft', _basic_uarray.fft, x, n, axis, norm,
+    return _execute_nD('fft', _basic_uarray.fft, x, n, axis, norm,
                        overwrite_x, workers, plan)
 
 
@@ -299,7 +282,7 @@ def ifft(x, n=None, axis=-1, norm=None, overwrite_x=False, workers=None, *,
     >>> plt.show()
 
     """
-    return _execute_1D('ifft', _basic_uarray.ifft, x, n, axis, norm,
+    return _execute_nD('ifft', _basic_uarray.ifft, x, n, axis, norm,
                        overwrite_x, workers, plan)
 
 
@@ -393,7 +376,7 @@ def rfft(x, n=None, axis=-1, norm=None, overwrite_x=False, workers=None, *,
     exploited to compute only the non-negative frequency terms.
 
     """
-    return _execute_1D('rfft', _basic_uarray.rfft, x, n, axis, norm,
+    return _execute_nD('rfft', _basic_uarray.rfft, x, n, axis, norm,
                        overwrite_x, workers, plan)
 
 
@@ -494,7 +477,7 @@ def irfft(x, n=None, axis=-1, norm=None, overwrite_x=False, workers=None, *,
     specified, and the output array is purely real.
 
     """
-    return _execute_1D('irfft', _basic_uarray.irfft, x, n, axis, norm,
+    return _execute_nD('irfft', _basic_uarray.irfft, x, n, axis, norm,
                        overwrite_x, workers, plan)
 
 
@@ -577,7 +560,7 @@ def hfft(x, n=None, axis=-1, norm=None, overwrite_x=False, workers=None, *,
     >>> hfft(signal, 10)  # Input entire signal and truncate
     array([  0.,   5.,   0.,  15.,  -0.,   0.,   0., -15.,  -0.,   5.])
     """
-    return _execute_1D('hfft', _basic_uarray.hfft, x, n, axis, norm,
+    return _execute_nD('hfft', _basic_uarray.hfft, x, n, axis, norm,
                        overwrite_x, workers, plan)
 
 
@@ -644,7 +627,7 @@ def ihfft(x, n=None, axis=-1, norm=None, overwrite_x=False, workers=None, *,
     >>> ihfft(spectrum)
     array([ 1.-0.j,  2.-0.j,  3.-0.j,  4.-0.j]) # may vary
     """
-    return _execute_1D('ihfft', _basic_uarray.ihfft, x, n, axis, norm,
+    return _execute_nD('ihfft', _basic_uarray.ihfft, x, n, axis, norm,
                        overwrite_x, workers, plan)
```

</details>

Reply:
That might be fine (and preferable). I don't think that `uarray` enforces using keywords, I just thought it may be best practice to use them.

Reply:
That LGTM too, seems like an improvement